### PR TITLE
Frozen wheel sessions

### PR DESCRIPTION
**Pull request by josefbitzenhofer**

### CODE REVIEW

File: viral/cache_2p_sessions.py [hunk 1]
```diff
@@ -147,9 +147,55 @@ def add_daq_times_to_trial(
...
+    ), "Last chunk length does not match expected length"
+
+    assert not np.any(
+        np.isin(behaviour_times, frame_times[0])
+    ), "Behavioral pulses detected in pre-training period!"
+
+    assert not np.any(
+        np.isin(behaviour_times, frame_times[-1])
+    ), "Behavioral pulses detected in post-training period!"
+    return first_chunk, last_chunk
```
Comment (neuromantic99) [inline viral/cache_2p_sessions.py:189]: This would only error if there's a frame at exactly the same time as a spacer, you need to check that there's no behaviour times at any point during the first and last chunk

File: viral/cache_2p_sessions.py [hunk 2]
```diff
@@ -411,60 +499,68 @@ def main() -> None:
...
-                    and "reversal learning" not in session_type
-                ):
-                    print(f"Skipping {mouse_name} {date} {session_type}")
-                    continue
-
-                if not row["Sync file"]:
-                    print(
-                        f"Skipping {mouse_name} {date} {session_type} as no sync file"
+                wheel_blocked = row["Wheel blocked?"].lower() == "yes"
+                if date == "2025-03-28":
```
Comment (neuromantic99) [inline viral/cache_2p_sessions.py:511]: don't think you want this line in main  :D

This is a prime candaidate for an 
'AHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHH' comment
because you won't miss it if you look at a diff

File: viral/cache_2p_sessions.py [hunk 3]
```diff
@@ -411,60 +499,68 @@ def main() -> None:
...
-                    "learning day" not in session_type
-                    and "reversal learning" not in session_type
-                ):
-                    print(f"Skipping {mouse_name} {date} {session_type}")
-                    continue
-
-                if not row["Sync file"]:
-                    print(
-                        f"Skipping {mouse_name} {date} {session_type} as no sync file"
+                wheel_blocked = row["Wheel blocked?"].lower() == "yes"
```
Comment (neuromantic99) [inline viral/cache_2p_sessions.py:510]: I think this will error on old mice where there's no wheel blocked column, I would check if the column exists

File: viral/cache_2p_sessions.py [hunk 4]
```diff
@@ -351,11 +416,13 @@ def check_timestamps(
     if not trial_is_imaged(trial):
         return
```
Comment (neuromantic99) [inline viral/cache_2p_sessions.py:418]: why's this assertion been removed?

File: viral/cache_2p_sessions.py [hunk 5]
```diff
@@ -368,11 +435,29 @@ def check_timestamps(
...
+            # i.e. one minute into the behaviour is at least 15 mins into the entire session
+            # TODO: why is the offset so big in some trials?
+            if trial.trial_start_time / 60 < 50:
+                assert (
+                    abs(offset) <= 0.01
+                ), "Tiff timestamp does not match daq timestamp"
+            else:
+                assert (
+                    abs(offset) <= 0.015
+                ), "Tiff timestamp does not match daq timestamp"
```
Comment (neuromantic99) [inline viral/cache_2p_sessions.py:460]: Very nice, but you can collapse this into a single if else with something like

```python
increase_offset_allowance_time = 30 if not wheel_blocked else 50
if trial.trial_start_time / 60 < increase_offset_allowance_time:
    bla bla bla
```

What does your TODO: why is the offset so big in some trials mean?

File: viral/single_session.py [hunk 6]
```diff
@@ -365,7 +365,9 @@ def plot_speed_reward_unrewarded(
     prev_unrewarded = []
 
     first_position = 0
-    last_position = 200
+    # TODO: change back!
+    # last_position = 200
+    last_position = 180
```
Comment (neuromantic99) [inline viral/single_session.py:370]: :D

File: viral/speed_related_cells.py [hunk 7]
```diff
@@ -56,26 +58,22 @@ def get_speed_pearsonr(
...
-            pearsonr(activity_speed[i, :], mean_speed_per_bin)[0]
+            correlation_func(activity_speed[i, :], mean_speed_per_bin)[0]
             for i in range(activity_speed.shape[0])
         ]
     )
 
 
+def get_speed_pearsonr(
+    activity_speed: np.ndarray, mean_speed_per_bin: np.ndarray
+) -> np.ndarray:
```
Comment (neuromantic99) [inline viral/speed_related_cells.py:69]: these functions probably aren't necessary

File: .gitignore [hunk 8]
```diff
@@ -176,4 +176,6 @@ cython_debug/
 #  be found at https://github.com/github/gitignore/blob/main/Global/JetBrains.gitignore
 #  and can be added to the global gitignore or merged into this file.  For a more nuclear
 #  option (not recommended) you can uncomment the following to ignore the entire idea folder.
-#.idea/
\ No newline at end of file
+#.idea/
+
+.vscode/settings.json
```
[the remark this answers is not in the document]
Comment (josefbitzenhofer) [reply .gitignore:181]: I don't have a strong opinion on this, whatever you prefer tbh

File: viral/cache_2p_sessions.py [hunk 9]
```diff
@@ -192,10 +238,59 @@ def add_imaging_info_to_trials(
...
+            == len(frame_times_daq)
+            - 2 * len(behavioral_chunk_lens_tiffs)
+            - excluded_frames
+        )
+    else:
+        assert (
+            len(valid_frame_times)
+            == sum(behavioral_chunk_lens_tiffs)
+            == len(frame_times_daq) - 2 * len(behavioral_chunk_lens_tiffs)
+        )
```
[the remark this answers is not in the document]
Comment (josefbitzenhofer) [reply viral/cache_2p_sessions.py:292]: Ahem, I don't have any excuse tbh, think I moved the logic back and forth but you're 100% right

File: viral/cache_2p_sessions.py (hunk 5 above)
Comment (josefbitzenhofer) [reply viral/cache_2p_sessions.py:460]: Is this what you had in mind?
Also concerning my todo comment: I had one session where the offset was too big to pass the assertion, so will keep an eye on it

File: viral/cache_2p_sessions.py (hunk 2 above)
Comment (josefbitzenhofer) [reply viral/cache_2p_sessions.py:511]: Nope, I don't want this in main

File: viral/single_session.py (hunk 6 above)
Comment (josefbitzenhofer) [reply viral/single_session.py:370]: :'(

File: viral/utils.py [hunk 10]
```diff
@@ -392,12 +392,26 @@ def shuffle(x: np.ndarray) -> np.ndarray:
...
     two different sampling rates: 1,000 and 10,000. The sessions should be between 30 and 100 minutes.
     """
-    if 30 < len(frame_clock) / 1000 / 60 < 100:
+    max_duration = 120 if wheel_blocked else 100
+    if 30 < len(frame_clock) / 1000 / 60 < max_duration:
         return 1000
-    elif 30 < len(frame_clock) / 10000 / 60 < 100:
+    elif 30 < len(frame_clock) / 10000 / 60 < max_duration:
         return 10000
     raise ValueError("Could not determine sampling rate")
```
[the remark this answers is not in the document]
Comment (josefbitzenhofer) [reply viral/utils.py:404]: Well, yeah :'(
It is just one more line, can I keep it? 👉🏻👈🏻

[the remark this answers is not in the document]
Comment (neuromantic99) [reply viral/utils.py:404]: "just one more line" add up ;)

File: viral/cache_2p_sessions.py (hunk 4 above)
Comment (josefbitzenhofer) [reply viral/cache_2p_sessions.py:418]: When I removed it, I was under the wrong impression that it only accounts for behaviour + imaging, but I put it back in here

File: viral/utils.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (josefbitzenhofer) [reply viral/utils.py:404]: https://media1.tenor.com/m/1NrJp6lHkh0AAAAd/baby-bat-mavis.gif

[the remark this answers is not in the document]
Comment (josefbitzenhofer) [reply viral/utils.py:404]: You'll get what you want :'(

File: tests/test_cache_2p_sessions.py [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+    behaviour_times = np.arange(27100, 36900) * (sampling_rate / 30)
+
+    expected = (0, 26999), (36999, 63999)
+    result = extract_frozen_wheel_chunks(
+        stack_lengths_tiffs=stack_lengths_tiff,
+        valid_frame_times=valid_frame_times,
+        behaviour_times=behaviour_times,
+        sampling_rate=sampling_rate,
+    )
+    assert expected == tuple(tuple(int(x) for x in t) for t in result)
```
Comment (neuromantic99) [inline tests/test_cache_2p_sessions.py:21]:
```suggestion
    assert expected == result
```
seems to work for me

File: viral/cache_2p_sessions.py [hunk 12]
```diff
@@ -147,9 +147,77 @@ def add_daq_times_to_trial(
...
+        sampling_rate (int):                        The sampling rate of the DAQ system in Hz.
+
+    Returns:
+        tuple[tuple[int, int], tuple[int, int]]:    First chunk and last chunk, with their respective start and end frame.
+    """
+
+    # first chunk (before behavioural chunks)
+    first_chunk_len = stack_lengths_tiffs[0]
+    first_chunk = (0, first_chunk_len - 1)  # start and end frame
+    first_chunk_times = valid_frame_times[first_chunk[0] : first_chunk[1] + 1]
```
Comment (neuromantic99) [inline viral/cache_2p_sessions.py:170]: Sorry I know this is gonna change all the code, but i think we shouldn't offset everything by 1.

Currently if you want to get the frozen wheel chunk from the result returned here you need to do

```python
dff[first_chunk[0]: first_chunk[1] + 1]
```

I would just stick to everything being 0 indexed. I know that you're correct currently e.g. the 36999th frame is the final one but i think this is easier going forward

File: viral/cache_2p_sessions.py [hunk 13]
```diff
@@ -216,28 +294,33 @@ def add_imaging_info_to_trials(
...
-
-    """
-
+    # Consistently, the number of triggers recorded is two more than the number of frames recorded (for the imaged behaviour chunks).
+    # This only occurs when the imaging is manually stopped before a grab is complete (confirmed by counting triggers
+    # from a completed grab).
+    # The reason for first extra frame is obvious (we stop the imaging mid-way through a frame so it is not saved).
+    # The second happens for unclear reasons but must be at the end as there are no extra frames in the middle and the first
+    # frame is relaibly correct
+    # Possible we may see a recording with one extra frame if the imaging is stopped on flyback. The error below will catch this.
```
Comment (neuromantic99) [inline viral/cache_2p_sessions.py:323]: can you put my docstring back :D

File: viral/cache_2p_sessions.py [hunk 14]
```diff
@@ -147,9 +147,77 @@ def add_daq_times_to_trial(
...
+
+    # We want to make sure the chunks are >= 15 mins but < 20 mins
+    # 15 mins = 27,000 frames
+    # 20 mins = 36,000 frames
+    assert (
+        27000 <= first_chunk_len < 36000
+    ), "First chunk length does not match expected length"
+    assert (
+        15 * 60 * sampling_rate
+        <= (last_chunk_times[-1] - last_chunk_times[0])
```
Comment (neuromantic99) [inline viral/cache_2p_sessions.py:189]:
```suggestion
        <= (first_chunk_times[-1] - first_chunk_times[0])
```

File: viral/imaging_utils.py [hunk 15]
```diff
@@ -292,3 +293,15 @@ def get_ITI_matrix(
...
+def compute_clock_offset(
+    epochs: np.ndarray,
+    frame_times_daq: np.ndarray,
+    daq_start_time: datetime,
+    sampling_rate: int,
+) -> datetime:
+    """Compute offset between the DAQ clock and imaging clock."""
+    return time_list_to_datetime(epochs[0]) - (
+        daq_start_time + timedelta(seconds=float(frame_times_daq[0] / sampling_rate))
+    )
```
[the remark this answers is not in the document]
Comment (josefbitzenhofer) [reply viral/imaging_utils.py:307]: This was a remnant of my approach of fixing the clocks being off due to timezone issues, I guess we won't need this anymore.

File: viral/cache_2p_sessions.py [hunk 16]
```diff
@@ -127,11 +134,13 @@ def add_daq_times_to_trial(
...
-            np.argmin(np.abs(valid_frame_times - state.start_time_daq))
+        state.closest_frame_start = (
+            int(np.argmin(np.abs(valid_frame_times - state.start_time_daq)))
+            + offset_after_pre_epoch
         )
-        state.closest_frame_end = int(
-            np.argmin(np.abs(valid_frame_times - state.end_time_daq))
+        state.closest_frame_end = (
+            int(np.argmin(np.abs(valid_frame_times - state.end_time_daq)))
+            + offset_after_pre_epoch
```
Comment (neuromantic99) [inline viral/cache_2p_sessions.py:143]: Can you add a check to test that neither the `closest_frame_start` or `closest_frame_end` are within the pre or post freeze periods?

File: viral/cache_2p_sessions.py [hunk 17]
```diff
@@ -142,16 +151,167 @@ def add_daq_times_to_trial(
 
     for event in trial.events_info:
         event.start_time_daq = float(bpod_to_daq(event.start_time))
-        event.closest_frame = int(
-            np.argmin(np.abs(valid_frame_times - event.start_time_daq))
+        event.closest_frame = (
+            int(np.argmin(np.abs(valid_frame_times - event.start_time_daq)))
+            + offset_after_pre_epoch
+        )
+
```
Comment (neuromantic99) [inline viral/cache_2p_sessions.py:158]: same 'not in the wheel freeze periods' check

File: tests/test_cache_2p_sessions.py [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+    ):
+        extract_frozen_wheel_chunks(
+            stack_lengths_tiffs=stack_lengths_tiff,
+            valid_frame_times=valid_frame_times,
+            behaviour_times=behaviour_times,
+            sampling_rate=sampling_rate,
+        )
+
+
+def test_extract_frozen_wheel_chunks_unexpected_chunk_len() -> None:
```
[the remark this answers is not in the document]
Comment (josefbitzenhofer) [reply tests/test_cache_2p_sessions.py:77]: added at least a simple one but w/o edge cases
